File: java/src/test/java/org/oojs/FailFastTest.java
package org.oojs;

import org.junit.jupiter.api.Test;
import org.oojs.model.Schema;

import java.util.List;

import static org.junit.jupiter.api.Assertions.*;

class FailFastTest extends TestHelpers {

    @Test void fail_fast_returns_single_error() {
        Object[] ar = animalRegistry();
        Registry r = (Registry) ar[0];
        Schema schema = (Schema) ar[1];
        List<ValidationError> errs = new Validator(r, true).validate(
                map("_type", "Dog"), schema.types.get("Animal"), schema);
        assertEquals(1, errs.size());
    }

    @Test void full_mode_returns_all_errors() {
        Object[] ar = animalRegistry();
        Registry r = (Registry) ar[0];
        Schema schema = (Schema) ar[1];
        List<ValidationError> errs = new Validator(r, false).validate(
                map("_type", "Dog"), schema.types.get("Animal"), schema);
        assertTrue(errs.size() >= 2);
    }
}

File: php/tests/FailFastTest.php
<?php

declare(strict_types=1);

namespace Oojs\Tests;

use PHPUnit\Framework\TestCase;

use function Oojs\validate;

/**
 * §9.6 — Fail-fast mode.
 */
class FailFastTest extends TestCase
{
    use TestHelpers;

    public function test_fail_fast_returns_single_error(): void
    {
        [$r, $schema] = self::animalRegistry();
        $instance = ['_type' => 'Dog']; // missing name and breed
        $errs = validate($instance, $schema->types['Animal'], $schema, $r, failFast: true);
        $this->assertCount(1, $errs);
    }

    public function test_full_mode_returns_all_errors(): void
    {
        [$r, $schema] = self::animalRegistry();
        $instance = ['_type' => 'Dog']; // missing name and breed → ≥2 required errors
        $errs = validate($instance, $schema->types['Animal'], $schema, $r, failFast: false);
        $this->assertGreaterThanOrEqual(2, count($errs));
    }
}
